lib2/provider/awssamlprovider.go

```diff
@@ -0,0 +1,450 @@
+		}
+	}
+
+	provider := AwsSamlProvider{
+		AwsSamlProviderOptions: opts,
+		oktaClient:             oktaClient,
+		keyring:                kr,
+		profileARN:             profileARN,
+		sessions:               sessions,
+		profile:                profile,
+	}
+
+	if region := opts.Profiles[source]["region"]; region != "" {
+		provider.awsRegion = region
+	}
+	err = provider.getSamlURL()
+	if err != nil {
+		return nil, err
+	}
+	return &provider, nil
+}
+
+// Gets a set of STS credentials to access AWS services.
+func (p *AwsSamlProvider) Retrieve() (credentials.Value, error) {
+
+	window := p.ExpiryWindow
+	if window == 0 {
+		window = time.Minute * 5
+	}
+
+	// TODO(nick): why are we using the source profile name and not the actual profile's name?
+	source := lib.SourceProfile(p.profile, p.Profiles)
+	profileConf, ok := p.Profiles[p.profile]
+	if !ok {
+		return credentials.Value{}, fmt.Errorf("missing profile named %s", p.profile)
+	}
+	key := sessioncache.OrigKey{
+		ProfileName: source,
+		ProfileConf: profileConf,
+		Duration:    p.SessionDuration,
+	}
+
+	var creds sts.Credentials
+	if cachedSession, err := p.sessions.Get(key); err != nil {
+		creds, err = p.getSamlSessionCreds()
+		if err != nil {
+			return credentials.Value{}, xerrors.Errorf("getting creds via SAML: %w", err)
+		}
+		newSession := sessioncache.Session{
+			Name:        p.roleSessionName(),
+			Credentials: creds,
+		}
+		if err = p.sessions.Put(key, &newSession); err != nil {
+			return credentials.Value{}, xerrors.Errorf("putting to sessioncache", err)
+		}
+
+		// TODO(nick): not really clear why this is done
+		p.defaultRoleSessionName = newSession.Name
+	} else {
+		creds = cachedSession.Credentials
+		p.defaultRoleSessionName = cachedSession.Name
+	}
+
+	log.Debugf("Using session %s, expires in %s",
+		(*(creds.AccessKeyId))[len(*(creds.AccessKeyId))-4:],
+		creds.Expiration.Sub(time.Now()).String())
+
+	// If SourceProfile returns the same source then we do not need to assume a
+	// second role. Not assuming a second role allows us to assume IDP enabled
+	// roles directly.
+	if p.profile != source {
+		if role, ok := p.Profiles[p.profile]["role_arn"]; ok {
+			var err error
+			creds, err = p.assumeRoleFromSession(creds, role)
+			if err != nil {
+				return credentials.Value{}, err
+			}
+
+			log.Debugf("using role %s expires in %s",
+				(*(creds.AccessKeyId))[len(*(creds.AccessKeyId))-4:],
+				creds.Expiration.Sub(time.Now()).String())
+		}
+	}
+
+	p.SetExpiration(*(creds.Expiration), window)
+	p.Expires = *(creds.Expiration)
+
+	value := credentials.Value{
+		AccessKeyID:     *(creds.AccessKeyId),
+		SecretAccessKey: *(creds.SecretAccessKey),
+		SessionToken:    *(creds.SessionToken),
+		ProviderName:    "okta",
+	}
+
+	return value, nil
+}
+
+// GetRoleARN uses temporary credentials to call AWS's get-caller-identity and
+// returns the assumed role's ARN
+func (p *AwsSamlProvider) GetRoleARNWithRegion(creds credentials.Value) (string, error) {
```

Comment:
Using these AWS structs feels very weird to me. IMO, we should make our own.

As an example, we take in a `credentials.Value` here, which has a `ProviderName` field. We don't use that field in this method.

Not a huge deal.

Reply:
I'd like to delete this method, it seems like more trouble than it's worth.

Any objection?

Reply:
Yep, if the CLI doesn't call it or you can simplify it somehow, by all means, go for it.